### ISSUE TITLE
Inconsistent networking issues with swarm in docker

### ISSUE THREAD

**Issue by duglin** [docker/docker#24684]
Using this script:
```
(docker rm -f manager node1 node2;docker network rm mynet;rm -f cmds)&>/dev/null

bold=$(tput bold)
normal=$(tput sgr0)

function doit() {
	set +e
	echo -n $bold"$ "
	if [[ "$SKIP" == "" ]]; then read -s -n 1 ; fi
	echo -n $*$normal" "
	echo "$*" >> cmds
	if [[ "$SKIP" == "" ]]; then read -s -n 1 ; fi
	echo
	$* 2>err >out
	rc=$?
	cat err out
	[[ "$rc" == "0" ]] || exit
	echo
	set -e
}

set -e

doit docker network create mynet
doit docker run --net mynet --privileged -dti --name manager --hostname manager ibmdojo/dind
doit docker run --net mynet --privileged -dti --name node1 --hostname node1 ibmdojo/dind
doit docker run --net mynet --privileged -dti --name node2 --hostname node2 ibmdojo/dind

doit docker exec -ti manager docker swarm init --secret please
doit docker exec -ti node1 docker swarm join --secret please manager
doit docker exec -ti node2 docker swarm join --secret please manager

doit docker exec -ti manager docker service create --name mini -p 80:80 ibmdojo/mini

doit docker exec -ti manager docker service ls
echo -e $bold"Waiting for it to start...\\n"$normal
while !(docker exec manager docker service ls|grep "1/"&>/dev/null);do sleep 1;done

doit docker exec -ti manager docker service ls
doit docker exec -ti manager curl manager
doit docker exec -ti manager curl node1
doit docker exec -ti manager curl node2

doit docker exec -ti manager docker ps
doit docker exec -ti node1 docker ps
doit docker exec -ti node2 docker ps
doit docker exec -ti manager docker service scale mini=3

echo -e $bold"Waiting for it to scale...\\n"$normal
while !(docker exec manager docker service ls|grep "3/"&>/dev/null);do sleep 1;done

doit docker exec -ti node1 docker ps
doit docker exec -ti manager curl manager
doit docker exec -ti manager curl manager
```

And running it with `SKIP=1 ./demo` or just `./demo` (and hitting a key after each prompt), this should create 3 containers running a ubuntu based DinD. Those 3 DinDs are then connected up in a swarm cluster. It then creates a service (mini) in that cluster which is just an httpd server that spits out some information about the interfaces in the container.

When it works each of the `curl manager` commands should show something like this:
```
<html><b>Hostname: 3c7067c2d8a7<br>
<br>Interfaces:</b><br>
<pre>127.0.0.1
10.255.0.7
10.255.0.6
172.19.0.3
</pre><br><b>Date: 2016-07-15 13:39:54.907084391 +0000 UTC</b>
```

However, I'm getting inconsistent results.

For about 2 days this script would fail on the curl with:
```
$ docker exec -ti manager curl manager 
curl: (7) Failed to connect to manager port 80: Connection refused
```
then suddenly it start to work for me. The only thing that could have changed was the network in which my laptop was running (home vs work).  But even though it was failing for me at work one day, the next day it started to work.

Other people who have tried this can't get it to work at all with running the main docker in a ubuntu VM on a mac (which is my env), but can get it work when running docker directly on the mac via the new built-in docker vm stuff (can't remember the name).

I'm using master for the main docker and the DinD docker (commit 56fb100140f96a010dcc834ed1d1626ffbc28970) and running Ubuntu 15.10 for the main docker host. The DinD image is using Ubuntu 15.04 as its base image.

Couple of other interesting things I've noticed:
- notice above that the curl is returning more than one 10.x address:
```
10.255.0.7
10.255.0.6
```
I'm not if that's correct or not, but it seemed odd.
- when I run this, even in an env where it all works, the curl's that are run after I scale the service up to 3 instances pauses for a long time before it returns the output. This makes me wonder if there's a timing issue here.

Anyway, net, I'm not 100% sure if there's something wrong with my script or swarm but I thought I'd open the issue in case this is a sign of something that needs to be tweaked in the swarm networking logic.

ping @mavenugo

**Comment by mavenugo** [comment]
@duglin thanks for the detailed bug report. I will look into this today and analyze the issue. But in general, we have an issue with dind as spotted by @tonistiigi with shared `/var/run/docker/netns`. I am not sure if it is impacting your case or not.

BTW, This is a basic scenario and I will be surprised if it is seen outside of the dind setup you have. Could you reproduce this issue in a non-dind environment ?

**Comment by duglin** [comment]
@mavenugo I haven't had a chance yet to spin up a set of VMs to test it w/o DinD but will try to do so today.

In the meantime, another interesting data point is that in an environment where it fails people are able to do: `docker exec -ti manager curl -v 172.19.0.3` and get the correct results.  It appears that 172.19.0.3 maps to a container with the name `mini.1.czwi2swa0yd0aljhb9b2vn9r6` which I assume means its the first replica of the service, but its not in the container inspect output, I only see that when I inspect the `docker_gwbridge` network.  Is that the way it should be?

And, in that network inspect output, under the `Container` section it shows just 1 of the 'mini' replicas, not 3, and an ingress:
```
 "Containers": {
            "ec38100aa017e86185bf3ffa384bbd31fe31a5187454217dfce4ead64c1d7730": {
                "Name": "gateway_ec38100aa017",
                "EndpointID": "7c6dff43d74f7383e785e65027a4ea61e7d62ad69b4fe338e91ea564a67a6363",
                "MacAddress": "02:42:ac:13:00:03",
                "IPv4Address": "172.19.0.3/16",
                "IPv6Address": ""
            },
            "ingress-sbox": {
                "Name": "gateway_ingress-sbox",
                "EndpointID": "1b2a9ceea716fffb5ef1752a29e0d26b0308293edbe8427a4896fd3a7a72cb10",
                "MacAddress": "02:42:ac:13:00:02",
                "IPv4Address": "172.19.0.2/16",
                "IPv6Address": ""
            }
        },
```
is that the way it should be or should all 3 "mini" containers be in there?
Note: `ec38100aa017e86185...` is the mini.1 container.

**Comment by mavenugo** [comment]
@duglin `docker network inspect` unfortunately displays ONLY those containers that are local to that node. The user must manually execute `docker network inspect` on all the nodes to get the list of all the containers that backs the service.

**Comment by duglin** [comment]
Some more interesting data...
@mavenugo try running this script:
```
(docker service rm $(docker service ls -q);docker exec -ti node1 docker swarm leave;docker swarm leave --force;docker rm -f node1)&>/dev/null

bold=$(tput bold)
normal=$(tput sgr0)

function doit() {
    set +e
    echo -n $bold"$ "
    if [[ "$SKIP" == "" ]]; then read -s -n 1 ; fi
    echo -n $*$normal" "
    echo "$*" >> cmds
    if [[ "$SKIP" == "" ]]; then read -s -n 1 ; fi
    echo
    $* 2>err >out
    rc=$?
    cat err out
    [[ "$rc" == "0" ]] || exit
    echo
    set -e
}

set -e

doit docker swarm init --secret please
joinIP=$(grep 2377 out)

doit docker run --privileged -dti --name node1 --hostname node1 ibmdojo/dind
doit docker exec -ti node1 docker swarm join --secret please $joinIP

doit docker service create --name mini -p 80:80 ibmdojo/mini

doit docker service ls
echo -e $bold"Waiting for it to start...\\n"$normal
while !(docker service ls|grep "1/"&>/dev/null);do sleep 1;done

doit curl 127.0.0.1

doit docker service rm mini
doit docker rm -f node1
doit docker swarm leave --force
```
This will turn the main docker engine into a manager, create a dind container as a worker and then deploy a service. Similar to the main issue's scenario except the manager isn't within a container. This should work - as expected.  

Now, what's interesting is that after I run this script by main usecase now suddenly works!

I had to reboot my system over the weekend and suddenly the usecase stopped working. After playing around I found that running the above script suddenly makes it work. I'm wondering if this is due to some networking config that's lingering around?  But, this would seem to indicate that the secret necessary to make my dind scenario work might be within the docker's control - or when it doesn't work, might be due to docker not setting up the network fully.  Dunno - but its interesting  :-)

I've asked some other folks on my team (who couldn't get the original scenario to work) to try running this new script first - but I won't find out their results until tomorrow given its late.

**Comment by AkihiroSuda** [comment]
@mavenugo 
> docker network inspect unfortunately displays ONLY those containers that are local to that node. The user must manually execute docker network inspect on all the nodes to get the list of all the containers that backs the service.

The documentation is misleading, can I open a PR to remove these sentences?
https://docs.docker.com/engine/reference/commandline/network_inspect/
> For networks backed by multi-host network driver, such as Overlay, this command also shows the container endpoints in other hosts in the cluster. These endpoints are represented as “ep-{endpoint-id}” in the output. 

Or do you have a plan to implement better `docker network inspect` in the short term?

**Comment by AkihiroSuda** [comment]
p.s. please let me know if there is other issue ticket for tracking the "docker network inspect unfortunately displays ONLY those containers that are local to that node." problem (cc @thaJeztah )

**Comment by thaJeztah** [comment]
@AkihiroSuda yes, please open a docs PR for that; possibly we need to distinguish between overlay networks managed by an external k/v store and "swarm mode" overlay networks.

There is another issue tracking the "network inspect" part; https://github.com/docker/docker/issues/24186

**Comment by AkihiroSuda** [comment]
@thaJeztah Thank you, opened https://github.com/docker/docker/pull/27526
